test/lit/passes/gufa-refs.wast

```diff
@@ -1087,14 +1083,14 @@
         (local.get $parent)
       )
     )
-    ;; A ref.func is cast to a struct type, and then we read from that. The cast
-    ;; will trap at runtime, of course; for here, we should not error and also
-    ;; we can optimize these to unreachables. atm we filter out trapping
-    ;; contents in ref.cast, but not br_on_cast, so test both.
+    ;; An unrelated type is cast to a struct type, and then we read from that.
+    ;; The cast will trap at runtime, of course; for here, we should not error
+    ;; and also we can optimize these to unreachables. atm we filter out
+    ;; trapping contents in ref.cast, but not br_on_cast, so test both.
     (drop
       (struct.get $parent 0
         (ref.cast null $parent
-          (ref.func $func)
+          (struct.new $unrelated)
```

Comment:
Ah, thanks, I didn't realize the spec was getting more strict on that. So now there are no cases that the optimizer can prove a cast will definitely fail, then? If so then there is some code we can remove in optimize-instructions? If that sounds right I can look into it.

Reply:
The optimizer can still prove that casting struct A to struct B can fail if neither A <: B nor B <: A, so I don't think there's any code to remove. I might have missed some that's more specific to func and extern, though.

Reply:
Oh, thanks, I see - so they must share a supertype, but they might still not have `given ref's type <: intended cast type`. So there are still cases we can prove will fail (but fewer now).

Reply:
(The spec could be tightened even more, though, and not allow such things?)

Reply:
No, that would break subtype substitutability. If we have `B <: A` and `C <: A`, then we want casting `A` -> `B` to be valid. But then due to subtype substitutability, we need to allow substituting `C` for `A`, so casting `C` -> `A` needs to be valid as well, even though it will always trap.

Reply:
Makes sense, thanks!